rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -0,0 +1,390 @@
+      num_bins: Number of hash bins. By default None, no hashing is performed.
+      name: Name to give to the layer.
+      **kwargs: Keyword arguments to construct a layer.
+
+    Input: a list of int tensors of shape `[batch_size, d1, ..., dm]`
+    Output: a single int tensor of shape `[batch_size, d1, ..., dm]`
+
+    Example:
+
+    If the layer receives two inputs, `a=[[1, 2]]` and `b=[[1, 3]]`,
+    and if depth is 2, then
+    the output will be a single integer tensor `[[i, j, k, l]]`, where:
+    i is the index of the category "a1=1 and b1=1"
+    j is the index of the category "a1=1 and b2=3"
+    k is the index of the category "a2=2 and b1=1"
+    l is the index of the category "a2=2 and b2=3"
+    """
+    pass
+
+`tf.keras.layers.CategoryEncoding`
+CategoryEncoding(PreprocessingLayer):
+"""This layer transforms categorical inputs from index space to category space.
+   If input is dense/sparse, then output is dense/sparse."""
+
+  def __init__(self, num_categories, mode="sum", axis=-1, name=None, **kwargs):
+    """Constructs a CategoryEncoding layer.
+    Args:
+      num_categories: Number of elements in the vocabulary.
+      mode: how to reduce a categorical input if multivalent, can be one of "sum",  
+          "mean", "binary", "tfidf". It can also be None if this is not a multivalent input,
+          and simply needs to convert input from index space to category space. "tfidf" is only
+          valid when adapt is called on this layer.
+      axis: the axis to reduce, by default will be the last axis, specially true 
+          for sequential feature columns.
+      name: Name to give to the layer.
+     **kwargs: Keyword arguments to construct a layer.
+
+    Input: a int tensor of shape `[batch_size, d1, ..., dm-1, dm]`
+    Output: a float tensor of shape `[batch_size, d1, ..., dm-1, num_categories]`
+    """
+    pass
+
+`tf.keras.layers.CategoryHashing`
+CategoryHashing(PreprocessingLayer):
+"""This layer transforms categorical inputs to hashed output.
+   If input is dense/sparse, then output is dense/sparse."""
+  def __init__(self, num_bins, name=None, **kwargs):
+    """Constructs a CategoryHashing layer.
+
+    Args:
+      num_bins: Number of hash bins.
+      name: Name to give to the layer.
+      **kwargs: Keyword arguments to construct a layer.
+
+    Input: a int tensor of shape `[batch_size, d1, ..., dm]`
+    Output: a int tensor of shape `[batch_size, d1, ..., dm]`
+    """
+    pass
+
+```
+
+We also propose a `to_sparse` op to convert dense tensors to sparse tensors given user specified ignore values. This op can be used in both `tf.data` or [TF Transform](https://www.tensorflow.org/tfx/transform/get_started). In previous feature column world, "" is ignored for dense string input and -1 is ignored for dense int input.
+
+```python
+`tf.to_sparse`
+def to_sparse(input, ignore_value):
+  """Convert dense/sparse tensor to sparse while dropping user specified values.
```

Comment:
What is the benefit of calling this API with a SparseTensor input?

Reply:
to allow users to filter specified values, e.g., if the original input is already sparse:
indices = [[0,0], [1, 0], [1,1]]
values = ['A', '', 'C']
the user can still filter '' from it,

Reply:
This filtering can be built out of existing operations. You can call, [tf.where](https://www.tensorflow.org/api_docs/python/tf/where?version=stable) on `values` and pass the result to [tf.sparse.retain](https://www.tensorflow.org/api_docs/python/tf/sparse/retain?version=stable), which is simple enough that I do not see the point of introducing syntactic sugar for that.

Reply:
Makes sense. Let's just extend the `tf.sparse.from_dense` op.